File: plugins/plugin-memory/src/evaluators/long-term-extraction.ts
import {
  type IAgentRuntime,
  type Memory,
  type Evaluator,
  logger,
  ModelType,
  composePromptFromState,
} from "@elizaos/core";
import { MemoryService } from "../services/memory-service";
import { LongTermMemoryCategory, type MemoryExtraction } from "../types/index";

/**
 * Template for extracting long-term memories using cognitive science memory types
 */
const extractionTemplate = `# Task: Extract Long-Term Memory (Strict Criteria)

You are analyzing a conversation to extract ONLY the most critical, persistent information about the user using cognitive science memory categories.

# Recent Messages
{{recentMessages}}

# Current Long-Term Memories
{{existingMemories}}

# Memory Categories (Based on Cognitive Science)

## 1. EPISODIC Memory
Personal experiences and specific events with temporal/spatial context.
**Examples:**
- "User completed migration project from MongoDB to PostgreSQL in Q2 2024"
- "User encountered authentication bug in production on March 15th"
- "User had a negative experience with Docker networking in previous job"

**Requirements:**
- Must include WHO did WHAT, WHEN/WHERE
- Must be a specific, concrete event (not a pattern)
- Must have significant impact or relevance to future work

## 2. SEMANTIC Memory
General facts, concepts, knowledge, and established truths about the user.
**Examples:**
- "User is a senior backend engineer with 8 years experience"
- "User specializes in distributed systems and microservices architecture"
- "User's primary programming language is TypeScript"
- "User works at Acme Corp as technical lead"

**Requirements:**
- Must be factual, timeless information
- Must be explicitly stated or demonstrated conclusively
- No speculation or inference from single instances
- Core identity, expertise, or knowledge only

## 3. PROCEDURAL Memory
Skills, workflows, methodologies, and how-to knowledge.
**Examples:**
- "User follows strict TDD workflow: write tests first, then implementation"
- "User prefers git rebase over merge to maintain linear history"
- "User's debugging process: check logs → reproduce locally → binary search"
- "User always writes JSDoc comments before implementing functions"

**Requirements:**
- Must describe HOW user does something
- Must be a repeated, consistent pattern (seen 3+ times or explicitly stated as standard practice)
- Must be a workflow, methodology, or skill application
- Not one-off preferences

# ULTRA-STRICT EXTRACTION CRITERIA

## ✅ DO EXTRACT (Only These):

**EPISODIC:**
- Significant completed projects or milestones
- Important bugs, incidents, or problems encountered
- Major decisions made with lasting impact
- Formative experiences that shape future work

**SEMANTIC:**
- Professional identity (role, title, company)
- Core expertise and specializations (stated explicitly or demonstrated conclusively)
- Primary languages, frameworks, or tools (not exploratory use)
- Established facts about their work context

**PROCEDURAL:**
- Consistent workflows demonstrated 3+ times or explicitly stated
- Standard practices user always follows
- Methodology preferences with clear rationale
- Debugging, testing, or development processes

## ❌ NEVER EXTRACT:

- **One-time requests or tasks** (e.g., "can you generate an image", "help me debug this")
- **Casual conversations** without lasting significance
- **Exploratory questions** (e.g., "how does X work?")
- **Temporary context** (current bug, today's task)
- **Preferences from single occurrence** (e.g., user asked for code once)
- **Social pleasantries** (thank you, greetings)
- **Testing or experimentation** (trying out a feature)
- **Common patterns everyone has** (likes clear explanations)
- **Situational information** (working on feature X today)
- **Opinions without persistence** (single complaint, isolated praise)
- **General knowledge** (not specific to user)

# Quality Gates (ALL Must Pass)

1. **Significance Test**: Will this matter in 3+ months?
2. **Specificity Test**: Is this concrete and actionable?
3. **Evidence Test**: Is there strong evidence (3+ instances OR explicit self-identification)?
4. **Uniqueness Test**: Is this specific to THIS user (not generic)?
5. **Confidence Test**: Confidence must be >= 0.85 (be VERY conservative)
6. **Non-Redundancy Test**: Does this add NEW information not in existing memories?

# Confidence Scoring (Be Conservative)

- **0.95-1.0**: User explicitly stated as core identity/practice AND demonstrated multiple times
- **0.85-0.94**: User explicitly stated OR consistently demonstrated 5+ times
- **0.75-0.84**: Strong pattern (3-4 instances) with supporting context
- **Below 0.75**: DO NOT EXTRACT (insufficient evidence)

# Critical Instructions

1. **Default to NOT extracting** - When in doubt, skip it
2. **Require overwhelming evidence** - One or two mentions is NOT enough
3. **Focus on what's PERSISTENT** - Not what's temporary or situational
4. **Verify against existing memories** - Don't duplicate or contradict
5. **Maximum 2-3 extractions per run** - Quality over quantity

**If there are no qualifying facts (which is common), respond with <memories></memories>**

# Response Format

<memories>
  <memory>
    <category>semantic</category>
    <content>User is a senior TypeScript developer with 8 years of backend experience</content>
    <confidence>0.95</confidence>
  </memory>
  <memory>
    <category>procedural</category>
    <content>User follows TDD workflow: writes tests before implementation, runs tests after each change</content>
    <confidence>0.88</confidence>
  </memory>
  <memory>
    <category>episodic</category>
    <content>User led database migration from MongoDB to PostgreSQL for payment system in Q2 2024</content>
    <confidence>0.92</confidence>
  </memory>
</memories>`;

/**
 * Parse XML memory extraction response
 */
function parseMemoryExtractionXML(xml: string): MemoryExtraction[] {
  const memoryMatches = xml.matchAll(
    /<memory>[\s\S]*?<category>(.*?)<\/category>[\s\S]*?<content>(.*?)<\/content>[\s\S]*?<confidence>(.*?)<\/confidence>[\s\S]*?<\/memory>/g
  );

  const extractions: MemoryExtraction[] = [];

  for (const match of memoryMatches) {
    const category = match[1].trim() as LongTermMemoryCategory;
    const content = match[2].trim();
    const confidence = parseFloat(match[3].trim());

    // Validate category
    if (!Object.values(LongTermMemoryCategory).includes(category)) {
      logger.warn(`Invalid memory category: ${category}`);
      continue;
    }

    if (content && !isNaN(confidence)) {
      extractions.push({ category, content, confidence });
    }
  }

  return extractions;
}

/**
 * Long-term Memory Extraction Evaluator
 *
 * Analyzes conversations to extract persistent facts about users that should be remembered
 * across all future conversations.
 */
export const longTermExtractionEvaluator: Evaluator = {
  name: "LONG_TERM_MEMORY_EXTRACTION",
  description: "Extracts long-term facts about users from conversations",
  similes: ["MEMORY_EXTRACTION", "FACT_LEARNING", "USER_PROFILING"],
  alwaysRun: true,

  validate: async (
    runtime: IAgentRuntime,
    message: Memory
  ): Promise<boolean> => {
    // Only run on user messages (not agent's own)
    if (message.entityId === runtime.agentId) {
      return false;
    }

    if (!message.content?.text) {
      return false;
    }

    const memoryService = runtime.getService("memory") as MemoryService | null;
    if (!memoryService) {
      return false;
    }

    const config = memoryService.getConfig();
    if (!config.longTermExtractionEnabled) {
      logger.debug("Long-term memory extraction is disabled");
      return false;
    }

    // Count total messages from this entity in this room
    const currentMessageCount = await runtime.countMemories(
      message.roomId,
      false,
      "messages"
    );

    const shouldRun = await memoryService.shouldRunExtraction(
      message.entityId,
      message.roomId,
      currentMessageCount
    );
    return shouldRun;
  },

  handler: async (runtime: IAgentRuntime, message: Memory): Promise<void> => {
    const memoryService = runtime.getService("memory") as MemoryService;
    if (!memoryService) {
      logger.error("MemoryService not found");
      return;
    }

    const config = memoryService.getConfig();
    const { entityId, roomId } = message;

    try {
      logger.info(`Extracting long-term memories for entity ${entityId}`);

      // Get recent conversation context
      const recentMessages = await runtime.getMemories({
        tableName: "messages",
        roomId,
        count: 20,
        unique: false,
      });

      const formattedMessages = recentMessages
        .sort((a, b) => (a.createdAt || 0) - (b.createdAt || 0))
        .map((msg) => {
          const sender =
            msg.entityId === runtime.agentId ? runtime.character.name : "User";
          return `${sender}: ${msg.content.text || "[non-text message]"}`;
        })
        .join("\n");

      // Get existing long-term memories
      const existingMemories = await memoryService.getLongTermMemories(
        entityId,
        undefined,
        30
      );
      const formattedExisting =
        existingMemories.length > 0
          ? existingMemories
              .map(
                (m) =>
                  `[${m.category}] ${m.content} (confidence: ${m.confidence})`
              )
              .join("\n")
          : "None yet";

      // Generate extraction using LLM
      const state = await runtime.composeState(message);
      const prompt = composePromptFromState({
        state: {
          ...state,
          recentMessages: formattedMessages,
          existingMemories: formattedExisting,
        },
        template: extractionTemplate,
      });

      const response = await runtime.useModel(ModelType.TEXT_LARGE, {
        prompt,
      });

      const extractions = parseMemoryExtractionXML(response);

      logger.info(`Extracted ${extractions.length} long-term memories`);

      // Store each extracted memory
      for (const extraction of extractions) {
        // Apply stricter confidence threshold (0.85 minimum)
        if (
          extraction.confidence >=
          Math.max(config.longTermConfidenceThreshold, 0.85)
        ) {
          await memoryService.storeLongTermMemory({
            agentId: runtime.agentId,
            entityId,
            category: extraction.category,
            content: extraction.content,
            confidence: extraction.confidence,
            source: "conversation",
            metadata: {
              roomId,
              extractedAt: new Date().toISOString(),
            },
          });

          logger.info(
            `Stored long-term memory: [${
              extraction.category
            }] ${extraction.content.substring(0, 50)}...`
          );
        } else {
          logger.debug(
            `Skipped low-confidence memory: ${
              extraction.content
            } (confidence: ${extraction.confidence}, threshold: ${Math.max(
              config.longTermConfidenceThreshold,
              0.85
            )})`
          );
        }
      }

      // Update the extraction checkpoint after successful extraction
      const currentMessageCount = await runtime.countMemories(
        roomId,
        false,
        "messages"
      );
      await memoryService.setLastExtractionCheckpoint(
        entityId,
        roomId,
        currentMessageCount
      );
      logger.debug(
        `Updated extraction checkpoint to ${currentMessageCount} for entity ${entityId} in room ${roomId}`
      );
    } catch (error) {
      logger.error({ error }, "Error during long-term memory extraction:");
    }
  },

  examples: [],
};

File: plugins/plugin-memory/rust/src/evaluators/long_term_extraction.rs
use async_trait::async_trait;
use tracing::{debug, info, warn};

use super::{EvaluatorContext, EvaluatorResult, MemoryEvaluator};
use crate::config::MemoryConfig;
use crate::types::{LongTermMemoryCategory, MemoryExtraction};

fn parse_memory_extraction_xml(xml: &str) -> Vec<MemoryExtraction> {
    let mut extractions = Vec::new();
    let mut remaining = xml;

    while let Some(mem_start) = remaining.find("<memory>") {
        if let Some(mem_end) = remaining[mem_start..].find("</memory>") {
            let memory_block = &remaining[mem_start..mem_start + mem_end + 9];

            let category = memory_block.find("<category>").and_then(|start| {
                memory_block[start + 10..]
                    .find("</category>")
                    .map(|end| memory_block[start + 10..start + 10 + end].trim())
            });

            let content = memory_block.find("<content>").and_then(|start| {
                memory_block[start + 9..]
                    .find("</content>")
                    .map(|end| memory_block[start + 9..start + 9 + end].trim().to_string())
            });

            let confidence = memory_block.find("<confidence>").and_then(|start| {
                memory_block[start + 12..]
                    .find("</confidence>")
                    .and_then(|end| {
                        memory_block[start + 12..start + 12 + end]
                            .trim()
                            .parse::<f64>()
                            .ok()
                    })
            });

            if let (Some(cat_str), Some(content), Some(conf)) = (category, content, confidence) {
                match cat_str.parse::<LongTermMemoryCategory>() {
                    Ok(category) => {
                        extractions.push(MemoryExtraction {
                            category,
                            content,
                            confidence: conf,
                            metadata: serde_json::json!({}),
                        });
                    }
                    Err(e) => {
                        warn!("Invalid memory category: {}", e);
                    }
                }
            }

            remaining = &remaining[mem_start + mem_end + 9..];
        } else {
            break;
        }
    }

    extractions
}

/// Evaluator for extracting long-term memories from conversations.
///
/// This evaluator analyzes conversation content to identify facts, preferences,
/// and other persistent information about users that should be stored in
/// long-term memory for future reference.
pub struct LongTermExtractionEvaluator {
    config: MemoryConfig,
}

impl LongTermExtractionEvaluator {
    /// Creates a new `LongTermExtractionEvaluator` with the given configuration.
    pub fn new(config: MemoryConfig) -> Self {
        Self { config }
    }

    /// Determines whether memory extraction should occur based on message counts.
    ///
    /// Returns `true` if the current message count has crossed an extraction
    /// checkpoint since the last extraction.
    pub fn should_extract(&self, current_count: i32, last_checkpoint: i32) -> bool {
        if !self.config.long_term_extraction_enabled {
            return false;
        }

        if current_count < self.config.long_term_extraction_threshold {
            return false;
        }

        let current_checkpoint = (current_count / self.config.long_term_extraction_interval)
            * self.config.long_term_extraction_interval;

        current_checkpoint > last_checkpoint
    }

    /// Parses an XML response into a list of memory extractions.
    pub fn parse_response(&self, xml: &str) -> Vec<MemoryExtraction> {
        parse_memory_extraction_xml(xml)
    }

    /// Filters extractions to only include those meeting confidence threshold.
    pub fn filter_by_confidence(
        &self,
        extractions: Vec<MemoryExtraction>,
    ) -> Vec<MemoryExtraction> {
        let threshold = self.config.long_term_confidence_threshold.max(0.85);
        extractions
            .into_iter()
            .filter(|e| e.confidence >= threshold)
            .collect()
    }
}

#[async_trait]
impl MemoryEvaluator for LongTermExtractionEvaluator {
    fn name(&self) -> &'static str {
        "LONG_TERM_MEMORY_EXTRACTION"
    }

    fn description(&self) -> &'static str {
        "Extracts long-term facts about users from conversations"
    }

    fn similes(&self) -> Vec<&'static str> {
        vec!["MEMORY_EXTRACTION", "FACT_LEARNING", "USER_PROFILING"]
    }

    async fn validate(&self, context: &EvaluatorContext) -> bool {
        if context.entity_id == context.agent_id {
            return false;
        }

        if context.message_text.is_empty() {
            return false;
        }

        let last_checkpoint = context
            .state
            .get("lastExtractionCheckpoint")
            .and_then(|v| v.as_i64())
            .map(|v| v as i32)
            .unwrap_or(0);

        self.should_extract(context.message_count, last_checkpoint)
    }

    async fn handler(&self, context: &EvaluatorContext) -> Option<EvaluatorResult> {
        info!(
            "Extracting long-term memories for entity {}",
            context.entity_id
        );

        debug!(
            "Long-term extraction evaluator triggered for entity {} with {} messages",
            context.entity_id, context.message_count
        );

        Some(EvaluatorResult {
            success: true,
            data: Some(serde_json::json!({
                "evaluator": self.name(),
                "entity_id": context.entity_id.to_string(),
                "room_id": context.room_id.to_string(),
                "message_count": context.message_count,
            })),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_memory_extraction_xml() {
        let xml = r#"
            <memories>
                <memory>
                    <category>semantic</category>
                    <content>User is a Rust developer</content>
                    <confidence>0.95</confidence>
                </memory>
                <memory>
                    <category>episodic</category>
                    <content>User mentioned working on a CLI tool</content>
                    <confidence>0.88</confidence>
                </memory>
            </memories>
        "#;

        let extractions = parse_memory_extraction_xml(xml);
        assert_eq!(extractions.len(), 2);

        assert_eq!(extractions[0].category, LongTermMemoryCategory::Semantic);
        assert_eq!(extractions[0].content, "User is a Rust developer");
        assert!((extractions[0].confidence - 0.95).abs() < f64::EPSILON);

        assert_eq!(extractions[1].category, LongTermMemoryCategory::Episodic);
        assert_eq!(
            extractions[1].content,
            "User mentioned working on a CLI tool"
        );
        assert!((extractions[1].confidence - 0.88).abs() < f64::EPSILON);
    }

    #[test]
    fn test_should_extract() {
        let config = MemoryConfig {
            long_term_extraction_enabled: true,
            long_term_extraction_threshold: 30,
            long_term_extraction_interval: 10,
            ..Default::default()
        };

        let evaluator = LongTermExtractionEvaluator::new(config);

        // Below threshold
        assert!(!evaluator.should_extract(20, 0));

        // At threshold, first extraction
        assert!(evaluator.should_extract(30, 0));

        // At threshold, already extracted
        assert!(!evaluator.should_extract(30, 30));

        // Next interval
        assert!(evaluator.should_extract(40, 30));
    }

    #[test]
    fn test_filter_by_confidence() {
        let config = MemoryConfig {
            long_term_confidence_threshold: 0.85,
            ..Default::default()
        };

        let evaluator = LongTermExtractionEvaluator::new(config);

        let extractions = vec![
            MemoryExtraction {
                category: LongTermMemoryCategory::Semantic,
                content: "High confidence".to_string(),
                confidence: 0.95,
                metadata: serde_json::json!({}),
            },
            MemoryExtraction {
                category: LongTermMemoryCategory::Semantic,
                content: "Low confidence".to_string(),
                confidence: 0.75,
                metadata: serde_json::json!({}),
            },
            MemoryExtraction {
                category: LongTermMemoryCategory::Episodic,
                content: "Threshold confidence".to_string(),
                confidence: 0.85,
                metadata: serde_json::json!({}),
            },
        ];

        let filtered = evaluator.filter_by_confidence(extractions);
        assert_eq!(filtered.len(), 2);
        assert_eq!(filtered[0].content, "High confidence");
        assert_eq!(filtered[1].content, "Threshold confidence");
    }
}
